File: CarbohydrateCalculatorJava/src/main/java/pages/CarbohydrateCalculatorPage.java
package pages;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.FindBy;
import org.openqa.selenium.support.PageFactory;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.Select;
import org.openqa.selenium.support.ui.WebDriverWait;
import utils.ConfigReader;

import java.time.Duration;
import java.util.List;
import java.util.stream.Collectors;

public class CarbohydrateCalculatorPage {

    private final WebDriver driver;
    private final WebDriverWait wait;

    // --- Locators ---

    @FindBy(id = "homelink")
    private WebElement logoLink;

    @FindBy(css = "a[href*='sign-in']")
    private WebElement signInLink;

    @FindBy(id = "cage")
    private WebElement ageInput;

    @FindBy(id = "csex1")
    private WebElement genderMale;

    @FindBy(id = "csex2")
    private WebElement genderFemale;

    @FindBy(id = "cheightfeet")
    private WebElement heightFeetInput;

    @FindBy(id = "cheightinch")
    private WebElement heightInchInput;

    @FindBy(id = "cpound")
    private WebElement weightInput;

    @FindBy(id = "cactivity")
    private WebElement activityDropdown;

    @FindBy(xpath = "//input[@value='Calculate']")
    private WebElement calculateButton;

    @FindBy(xpath = "//input[@value='Clear']")
    private WebElement clearButton;

    @FindBy(css = "div.bigtext")
    private WebElement resultSection;

    @FindBy(css = "font[color='red']")
    private WebElement errorBanner;

    @FindBy(xpath = "//a[contains(text(),'Settings')]")
    private WebElement settingsLink;

    @FindBy(id = "cbodyfat")
    private WebElement bodyFatInput;

    @FindBy(xpath = "//a[contains(text(),'Print')]")
    private WebElement printLink;

    @FindBy(id = "calcSearchTerm")
    private WebElement searchInput;

    @FindBy(xpath = "//a[contains(text(),'Other Units')]")
    private WebElement otherUnitsHeader;

    @FindBy(css = "table.cinfoT")
    private WebElement resultTable;

    public CarbohydrateCalculatorPage(WebDriver driver) {
        this.driver = driver;
        this.wait = new WebDriverWait(driver,
                Duration.ofSeconds(ConfigReader.getExplicitWait()));
        PageFactory.initElements(driver, this);
    }

    // --- Actions ---

    public void navigateTo() {
        driver.get(ConfigReader.getBaseUrl());
    }

    public void clickLogo() {
        wait.until(ExpectedConditions.elementToBeClickable(logoLink)).click();
    }

    public void clickSignIn() {
        wait.until(ExpectedConditions.elementToBeClickable(signInLink)).click();
    }

    public void enterAge(String age) {
        ageInput.clear();
        ageInput.sendKeys(age);
    }

    public void selectGender(String gender) {
        if (gender.equalsIgnoreCase("Male")) {
            genderMale.click();
        } else {
            genderFemale.click();
        }
    }

    public void enterHeight(String feet, String inches) {
        heightFeetInput.clear();
        heightFeetInput.sendKeys(feet);
        heightInchInput.clear();
        heightInchInput.sendKeys(inches);
    }

    public void enterWeight(String weight) {
        weightInput.clear();
        weightInput.sendKeys(weight);
    }

    public void selectActivity(String activityText) {
        Select select = new Select(activityDropdown);
        select.selectByVisibleText(activityText);
    }

    public String getSelectedActivity() {
        Select select = new Select(activityDropdown);
        return select.getFirstSelectedOption().getText();
    }

    public List<String> getActivityOptions() {
        Select select = new Select(activityDropdown);
        return select.getOptions().stream()
                .map(WebElement::getText)
                .collect(Collectors.toList());
    }

    public void clickCalculate() {
        calculateButton.click();
    }

    public void clickClear() {
        clearButton.click();
    }

    public void clickSettings() {
        settingsLink.click();
    }

    public void clickPrint() {
        printLink.click();
    }

    public void clickOtherUnits() {
        otherUnitsHeader.click();
    }

    public void enterSearchTerm(String term) {
        searchInput.clear();
        searchInput.sendKeys(term);
    }

    public boolean isResultDisplayed() {
        try {
            return resultSection.isDisplayed();
        } catch (Exception e) {
            return false;
        }
    }

    public boolean isResultTableDisplayed() {
        try {
            return resultTable.isDisplayed();
        } catch (Exception e) {
            return false;
        }
    }

    public String getErrorMessage() {
        try {
            return wait.until(ExpectedConditions.visibilityOf(errorBanner)).getText();
        } catch (Exception e) {
            return "";
        }
    }

    public String getAgeValue() {
        return ageInput.getAttribute("value");
    }

    public String getWeightValue() {
        return weightInput.getAttribute("value");
    }

    public boolean isBodyFatSectionVisible() {
        try {
            return wait.until(ExpectedConditions.visibilityOf(bodyFatInput)).isDisplayed();
        } catch (Exception e) {
            return false;
        }
    }

    public String getCurrentUrl() {
        return driver.getCurrentUrl();
    }

    public String getPageSource() {
        return driver.getPageSource();
    }

    public void fillDefaultData() {
        selectGender("Male");
        enterAge("25");
        enterHeight("5", "10");
        enterWeight("160");
        selectActivity("Moderate: exercise 4-5 times/week");
    }
}

File: CarbohydrateCalculatorTests/Pages/CarbohydrateCalculatorPage.cs
using OpenQA.Selenium;
using OpenQA.Selenium.Support.UI;

namespace CarbohydrateCalculatorTests.Pages
{
    public class CarbohydrateCalculatorPage
    {
        private readonly IWebDriver _driver;
        private readonly WebDriverWait _wait;

        private const string PageUrl = "https://www.calculator.net/carbohydrate-calculator.html";

        // Locators
        private readonly By _logoLink = By.CssSelector("#logo a");
        private readonly By _signInLink = By.CssSelector("a[href*='sign-in']");
        private readonly By _ageInput = By.Id("cage");
        private readonly By _genderMale = By.Id("csex1");
        private readonly By _genderFemale = By.Id("csex2");
        private readonly By _heightFeetInput = By.Id("cheightfeet");
        private readonly By _heightInchInput = By.Id("cheightinch");
        private readonly By _weightInput = By.Id("cpound");
        private readonly By _heightMetricInput = By.Id("cheightmeter");
        private readonly By _weightMetricInput = By.Id("ckg");
        private readonly By _activityDropdown = By.Id("cactivity");
        private readonly By _calculateButton = By.XPath("//input[@value='Calculate']");
        private readonly By _clearButton = By.XPath("//input[@value='Clear']");
        private readonly By _resultSection = By.CssSelector("table.cinfoT");
        private readonly By _errorBanner = By.CssSelector("font[color='red']");
        private readonly By _settingsLink = By.XPath("//a[contains(text(),'Settings')]");
        private readonly By _bodyFatInput = By.Name("cfatpct");
        private readonly By _printLink = By.XPath("//a[contains(text(),'Print')]");
        private readonly By _saveLink = By.XPath("//a[contains(text(),'Save')]");
        private readonly By _searchInput = By.Id("calcSearchTerm");
        private readonly By _otherUnitsHeader = By.XPath("//a[contains(text(),'Other Units')]");
        private readonly By _convertFromInput = By.Id("fromVal");
        private readonly By _resultTable = By.CssSelector("table.cinfoT");

        public CarbohydrateCalculatorPage(IWebDriver driver)
        {
            _driver = driver;
            _wait = new WebDriverWait(driver, TimeSpan.FromSeconds(10));
        }

        public void NavigateTo()
        {
            _driver.Navigate().GoToUrl(PageUrl);
        }

        public void ClickLogo()
        {
            ClickElement(_logoLink);
        }

        public void ClickSignIn()
        {
            ClickElement(_signInLink);
        }

        public void EnterAge(string age)
        {
            SetInputValue(_ageInput, age);
        }

        public void SelectGender(string gender)
        {
            if (gender.Equals("Male", StringComparison.OrdinalIgnoreCase))
                ClickElement(_genderMale);
            else
                ClickElement(_genderFemale);
        }

        public void EnterHeight(string feet, string inches)
        {
            SetInputValue(_heightFeetInput, feet);
            SetInputValue(_heightInchInput, inches);
        }

        public void EnterWeight(string weight)
        {
            SetInputValue(_weightInput, weight);
        }

        private void SetInputValue(By locator, string value)
        {
            var js = (IJavaScriptExecutor)_driver;
            var element = _driver.FindElement(locator);
            js.ExecuteScript("arguments[0].scrollIntoView(true);", element);
            js.ExecuteScript("arguments[0].value = '';", element);
            js.ExecuteScript("arguments[0].value = arguments[1];", element, value);
        }

        private void ClickElement(By locator)
        {
            var js = (IJavaScriptExecutor)_driver;
            var element = _driver.FindElement(locator);
            js.ExecuteScript("arguments[0].scrollIntoView(true);", element);
            js.ExecuteScript("arguments[0].click();", element);
        }

        public void SelectActivity(string activity)
        {
            var dropdown = new SelectElement(_driver.FindElement(_activityDropdown));
            dropdown.SelectByText(activity);
        }

        public string GetSelectedActivity()
        {
            var dropdown = new SelectElement(_driver.FindElement(_activityDropdown));
            return dropdown.SelectedOption.Text;
        }

        public List<string> GetActivityOptions()
        {
            var dropdown = new SelectElement(_driver.FindElement(_activityDropdown));
            return dropdown.Options.Select(o => o.Text).ToList();
        }

        public void ClickCalculate()
        {
            ClickElement(_calculateButton);
        }

        public void ClickClear()
        {
            ClickElement(_clearButton);
        }

        public void ClickSettings()
        {
            ClickElement(_settingsLink);
        }

        public void ClickPrint()
        {
            ClickElement(_printLink);
        }

        public void ClickOtherUnits()
        {
            ClickElement(_otherUnitsHeader);
        }

        public void EnterConvertFromValue(string value)
        {
            var element = _driver.FindElement(_convertFromInput);
            element.Clear();
            element.SendKeys(value);
        }

        public void EnterSearchTerm(string term)
        {
            var element = _driver.FindElement(_searchInput);
            element.Clear();
            element.SendKeys(term);
        }

        public bool IsResultDisplayed()
        {
            try
            {
                _wait.Until(d =>
                {
                    try { return d.FindElement(_resultSection).Displayed; }
                    catch { return false; }
                });
                return true;
            }
            catch
            {
                return false;
            }
        }

        public bool IsResultTableDisplayed()
        {
            try
            {
                return _driver.FindElement(_resultTable).Displayed;
            }
            catch (NoSuchElementException)
            {
                return false;
            }
        }

        public string GetErrorMessage()
        {
            try
            {
                return _driver.FindElement(_errorBanner).Text;
            }
            catch (NoSuchElementException)
            {
                return string.Empty;
            }
        }

        public string GetAgeValue()
        {
            return _driver.FindElement(_ageInput).GetAttribute("value");
        }

        public string GetWeightValue()
        {
            return _driver.FindElement(_weightMetricInput).GetAttribute("value");
        }

        public bool IsBodyFatSectionVisible()
        {
            try
            {
                _wait.Until(d => d.FindElement(_bodyFatInput).Displayed);
                return true;
            }
            catch
            {
                return false;
            }
        }

        public string GetCurrentUrl()
        {
            return _driver.Url;
        }

        public void EnterXssPayload(string payload)
        {
            SetInputValue(_weightMetricInput, payload);
        }

        public string GetPageSource()
        {
            return _driver.PageSource;
        }

        public bool IsConverterIframePresent()
        {
            try
            {
                var iframe = _driver.FindElement(By.CssSelector("iframe[src*='converter']"));
                return iframe.Displayed;
            }
            catch (NoSuchElementException)
            {
                return false;
            }
        }

        public void FillDefaultData(string age = "25", string heightCm = "180",
            string weightKg = "60", string gender = "Male", string activity = "Moderate: exercise 4-5 times/week")
        {
            SelectGender(gender);
            EnterAge(age);
            SetInputValue(_heightMetricInput, heightCm);
            SetInputValue(_weightMetricInput, weightKg);
            SelectActivity(activity);
        }
    }
}
